gotsh/gotsh.1
.\"
.\" Copyright (c) 2022 Stefan Sperling
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate$
.Dt GOTSH 1
.Os
.Sh NAME
.Nm gotsh
.Nd Game of Trees Shell
.Sh SYNOPSIS
.Nm Fl c Sq Cm git-receive-pack Ar repository-path
.Nm Fl c Sq Cm git-upload-pack Ar repository-path
.Sh DESCRIPTION
.Nm
is the network-facing interface to
.Xr gotd 8 .
It implements the server-side part of the Git network protocol used by
.Xr git 1
and
.Xr got 1 .
.Pp
.Nm
is not an interactive shell.
.Nm
is intended to be configured as the login shell of Git repository
user accounts on servers running
.Xr gotd 8 .
If users require a different login shell,
.Nm
can be installed in the command search path under the names
.Cm git-receive-pack
and
.Cm git-upload-pack ,
or
.Xr gitwrapper 1
can be used to select the appropriate command to run automatically.
.Pp
The users can then interact with
.Xr gotd 8
over the network.
When users invoke commands such as
.Cm got send
and
.Cm got fetch
on client machines,
.Xr got 1
will connect to the server with
.Xr ssh 1 .
.Nm
will facilitate communication between
.Xr gotd 8
running on the server machine and the
.Xr got 1
or
.Xr git 1
program running on the client machine.
.Pp
Users running
.Nm
should not have access to Git repositories by means other than
accessing the unix socket of
.Xr gotd 8
via
.Nm .
.Pp
It is recommended to restrict
.Xr ssh 1
features available to users of
.Nm .
See the
.Sx EXAMPLES
section for details.
.Sh ENVIRONMENT
.Bl -tag -width GOTD_UNIX_SOCKET
.It Ev GOTD_UNIX_SOCKET
Set the path to the unix socket which
.Xr gotd 8
is listening on.
If not specified, the default path
.Pa /var/run/gotd.sock
will be used.
.El
.Sh EXAMPLES
.Xr sshd_config 5
directives such as the following are recommended to protect the server
machine and any systems reachable from it, especially if anonymous users
are allowed to connect:
.Bd -literal -offset indent
Match User developer
    DisableForwarding yes
    PermitTTY no
.Ed
.Pp
It can be convenient to add all relevant users to a common group, such as
.Dq developers ,
and then use this group as the Match criteria:
.Bd -literal -offset indent
Match Group developers
    DisableForwarding yes
    PermitTTY no
.Ed
.Pp
Anonymous users can be given public read-only access by using a
.Xr gotd.conf 5
access rule such as the following:
.Bd -literal -offset indent
repository "public" {
	path "/var/git/public.git"
	permit ro anonymous
}
.Ed
.Pp
The anonymous user account should have a publicly known password, or can be
set up with an empty password in which case the user's
.Xr vipw 8
entry would look similar to this example:
.Bd -literal
anonymous::1002:1002::0:0:Anonymous:/home/anonymous:/usr/local/bin/gotsh
.Ed
.Pp
Use of an empty password must be explicitly allowed in
.Xr sshd_config 5 :
.Bd -literal -offset indent
Match User anonymous
    PasswordAuthentication yes
    PermitEmptyPasswords yes
    DisableForwarding yes
    PermitTTY no
.Ed
.Sh SEE ALSO
.Xr gitwrapper 1 ,
.Xr got 1 ,
.Xr ssh 1 ,
.Xr gotd.conf 5 ,
.Xr sshd_config 5 ,
.Xr gotd 8
.Sh AUTHORS
.An Stefan Sperling Aq Mt stsp@openbsd.org


CHANGES
* got 0.87; 2023-04-19
  see git repository history for per-change authorship information
- add gitwrapper(1)
- tog: resume blame and diff search from the first line
- fix crash in got log due to NULL-deref in got_object_blob_close
- add support for protecting references against 'got send -f' to gotd
- fix spurious empty packfile error from gotd when rewinding a branch
- tog: implement automated test harness
- update the base commit ID of unmodified files if the blob ID matches
- fix rebase/histedit -a leaving some files on the temporary branch
- make 'got revert' and 'got rm' work on non-existent directories
- got: flush stdout before printing the error in main()
- when aborting rebase/histedit/merge, unlink files added by merged changes
- fix 'got commit' using a bad parent commit ID when worktree is out-of-date
- allow no-op merge commits to be created
- fix sending merge commits
- show how to fetch a pull request in got.1 pull request example section

* got 0.86; 2023-03-12
  see git repository history for per-change authorship information
- fix race condition on NFS where log-message file's modify time may change
- fix gotd sending too large pack files in some cases
- support histedit fold operations which delete a file and then add it again
- make diffing files which changed into dirs and vice-versa possible
- handle files changing into directories during 'got update'
- add quoting to repository path sent to server for git-shell compatibility
- gotwebd: handle short reads and timeouts
- gotwebd: provide gotweb_render_page() entrypoint for all pages
- gotwebd: reply with non-200 HTTP status code on error

* got 0.85; 2023-03-07
  see git repository history for per-change authorship information
- gotwebd: add missing colon in diff view (patch by Josiah Frentsos)
- more preparation for eventual sha256 object ID support
- add test coverage for more tree conflict cases during merges
- fall back to vi(1) instead of ed(1) if neither EDITOR nor VISUAL are set
- in got.1, clarify what users are expected to do during 'histedit -e'
- gotd requires a config file; don't fail silently when it cannot be read
- regress: replace unportable ln -h option with rm && ln
- regress: make cmdline tests POSIX /bin/sh compatible
- gotd: remove more (all?) double process names in log
- don't pass -d to yacc during the build (patch by Josiah Frentsos)
- regress: override locale settings to force the "C" locale
- regress: replace "sed -i" with ed(1) for portable in-place editing
- fix gotd sometimes reading reused deltas from wrong pack file

* got 0.84; 2023-02-22
  too many changes to list all here; see git repository history for more;
  and see git repository history for per-change authorship information
- add 'got histedit -d' flag to drop all commits
- show worktree UUID in backout/cherrypick -l output
- several changes in preparation for eventual SHA256 object ID support
- make 'got rebase' work when the to-be-rebased branch has no parent commit
- fix bad line-wrapping in tog ref and tree views
- add horizontal scrolling support to tog ref and tree views
- create .pack and .idx files with filemode 0444, like Git does
- make 'got fetch' fetch the work tree's current branch as an implicit fallback
- improve 'got fetch' behaviour when work tree's branch is not on server
- gotwebd: fix briefs/tags navigation overlap
- drop double process name from some gotd logs
- tog: fix high CPU usage issue after starting search (reported by Mikhail)
- fix gotd exiting with abnormal error during client connection teardown
- fix gotd segfault in libevent while disconnecting clients
- tweak 'got commit' log-message validation: now checks timestamp and size > 0
- gotwebd: abort blame if the client disconnects midway through
- make 'got fetch -b branch' only fetch the named branch
- got/tog/gotadmin: call pledge(2) earlier where possible
- no longer error out if redundant "got commit -A $GOT_AUTHOR" option is used
- add 'gotadmin pack' -D flag to force generation of ref-deltas in pack files
- make 'got fetch' update cached HEAD symref if it has changed in remote repo
- add commit/histedit/merge/rebase -C option to commit unresolved conflicts
- make 'got status' check for merge conflict markers on newly added lines only
- fix read/write out of bounds in gitconfig file parser (reported by James Cook)
- fix "got fetch" hanging if remote repo is out-of-date (reported by James Cook)
- ignore patterns with trailing "/" now match directories (suggested by Lucas)
- make 'got merge' honour author setting in gitconfig (reported by James Cook)
- sync default values shown in gotwebd.conf(5) EXAMPLES section
- fix parsing of indented comments in gitconfig file (reported by James Cook)
- fix tog diff between arbitrary commits; regression from 0.80

* got 0.83; 2023-01-30
- fix usage display error in got merge command (patch by Mikhail)
- fix missing commits in pack files created with packed object enumeration
- avoid traversing enumerated commits more than once in got-read-pack
- fix ulimit-related test failures on sparc64 (tracey)
- got commit editor now shows log messages from backout and cherrypick (jamsek)
- new cherrypick/backout -l option to show recorded log messages (jamsek)
- new cherrypick/backout -X option to discard recorded log messages (jamsek)
- gotd: implement the delete-refs capability (op)
- fix histedit -m on a commit which only changes filemode bits
- gotsh.1: show how to set up anonymous public read-only repository access

* got 0.82; 2023-01-23
- fix comparison in tree object parser always evaluating to false (jamsek)
- add missing bounds-check in gitproto ref-line tokenizer
- gotd.8 and gotwebd.8 man page improvements (patch by Josiah Frentsos)
- make gotd session process accept just one flush packet at a time
- sort ENVIRONMENT entries in got.1 and tog.1 (op)
- only forward implicit flush packets from gotsh if they are expected
- return GOT_ERR_EOF from pkt.c if a read attempt indicates EOF

gotsh/gotsh.c
/*
 * Copyright (c) 2022 Stefan Sperling <stsp@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/queue.h>
#include <sys/socket.h>
#include <sys/un.h>

#include <err.h>
#include <event.h>
#include <imsg.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "got_error.h"
#include "got_serve.h"
#include "got_path.h"

#include "got_lib_dial.h"

#include "gotd.h"

static int chattygot;

__dead static void
usage(void)
{
	fprintf(stderr, "usage: %s -c '%s|%s repository-path'\n",
	    getprogname(), GOT_DIAL_CMD_SEND, GOT_DIAL_CMD_FETCH);
	exit(1);
}

static const struct got_error *
apply_unveil(const char *unix_socket_path)
{
#ifdef PROFILE
	if (unveil("gmon.out", "rwc") != 0)
		return got_error_from_errno2("unveil", "gmon.out");
#endif
	if (unveil(unix_socket_path, "w") != 0)
		return got_error_from_errno2("unveil", unix_socket_path);

	if (unveil(NULL, NULL) != 0)
		return got_error_from_errno("unveil");

	return NULL;
}

int
main(int argc, char *argv[])
{
	const struct got_error *error;
	const char *unix_socket_path;
	int gotd_sock = -1;
	struct sockaddr_un	 sun;
	char *gitcmd = NULL, *command = NULL, *repo_path = NULL;

#ifndef PROFILE
	if (pledge("stdio recvfd unix unveil", NULL) == -1)
		err(1, "pledge");
#endif

	unix_socket_path = getenv("GOTD_UNIX_SOCKET");
	if (unix_socket_path == NULL)
		unix_socket_path = GOTD_UNIX_SOCKET;

	error = apply_unveil(unix_socket_path);
	if (error)
		goto done;

	if (strcmp(argv[0], GOT_DIAL_CMD_SEND) == 0 ||
	    strcmp(argv[0], GOT_DIAL_CMD_FETCH) == 0) {
		if (argc != 2)
			usage();
		if (asprintf(&gitcmd, "%s %s", argv[0], argv[1]) == -1)
			err(1, "asprintf");
		error = got_dial_parse_command(&command, &repo_path, gitcmd);
	} else {
		if (argc != 3 || strcmp(argv[1], "-c") != 0)
			usage();
		error = got_dial_parse_command(&command, &repo_path, argv[2]);
	}
	if (error && error->code == GOT_ERR_BAD_PACKET)
		usage();
	if (error)
		goto done;

	if ((gotd_sock = socket(AF_UNIX, SOCK_STREAM, 0)) == -1)
		err(1, "socket");

	memset(&sun, 0, sizeof(sun));
	sun.sun_family = AF_UNIX;
	if (strlcpy(sun.sun_path, unix_socket_path,
	    sizeof(sun.sun_path)) >= sizeof(sun.sun_path))
		errx(1, "gotd socket path too long");
	if (connect(gotd_sock, (struct sockaddr *)&sun, sizeof(sun)) == -1)
		err(1, "connect: %s", unix_socket_path);

#ifndef PROFILE
	if (pledge("stdio recvfd", NULL) == -1)
		err(1, "pledge");
#endif
	error = got_serve(STDIN_FILENO, STDOUT_FILENO, command, repo_path,
	    gotd_sock, chattygot);
done:
	free(gitcmd);
	free(command);
	free(repo_path);
	if (gotd_sock != -1)
		close(gotd_sock);
	if (error) {
		fprintf(stderr, "%s: %s\n", getprogname(), error->msg);
		return 1;
	}

	return 0;
}
